api/src/main/java/org/apache/iceberg/catalog/TransactionalCatalog.java

```diff
@@ -0,0 +1,107 @@
+import org.apache.iceberg.exceptions.CommitFailedException;
+
+/**
+ * A {@link Catalog} that applies all mutations within a single transaction.
+ *
+ * <p>A TransactionalCatalog can spawn child transactions for multiple operations on different
+ * tables. All operations will be done within the context of a single Catalog-level transaction
+ * and they will either all be successful or all fail.
+ *
+ * <p>A TransactionalCatalog is initially active upon creation and will remain so until one of
+ * the following terminal actions occurs:
+ * <ul>
+ * <li>{@link rollback} is called.
+ * <li>{@link commit} is called.
+ * <li>The transaction expires while using Pessimistic {@link LockingMode}.
+ * <li>The transaction is terminated externally (for example, when a locking arbitrator
+ *     determines a deadlock between two transactions has occurred).
+ * <li>The underlying implementation determines that the transaction can no longer complete
+ *     successfully.
+ * </ul>
+ *
+ * <p>When one of the items above occurs, the transaction is no longer valid. Further use
+ * of the transaction will result in a {@link IllegalStateException} being thrown.
+ *
+ * <p>Nested transactions such as creating a new table may fail. Those failures alone do
+ * not necessarily result in a failure of the catalog-level transaction.
+ *
+ */
+public interface TransactionalCatalog extends Catalog, AutoCloseable {
+
+  /**
+   * An internal identifier associated with this transaction.
+   * @return An internal identifier.
+   */
+  String transactionId();
+
+  /**
+   * Return the current {@code IsolationLevel} for this transaction.
+   * @return The IsolationLevel for this transaction.
+   */
+  IsolationLevel isolationLevel();
+
+  /**
+   * Return the {@link LockingMode} for this transaction.
+   * @return The LockingMode for this transaction.
+   */
+  LockingMode lockingMode();
+
+  /**
+   * Whether the current transaction is still active/open.
+   * @return True until a terminal action occurs.
+   */
+  boolean active();
+
+  /**
+   * Aborts the set of operations here and makes this TransactionalCatalog inoperable.
+   *
+   * <p>Once called, no further operations can be done against this catalog. If any
+   * operations are attempted, {@link IllegalStateException} will be thrown.
+   */
+  void rollback();
+
+  /**
+   * Commit the pending changes from all nested transactions against the Catalog.
+   *
+   * <p>Once called, no further operations can be done against this catalog. If any
+   * operations are attempted, {@link IllegalStateException} will be thrown.
+   *
+   * @throws CommitFailedException If the updates cannot be committed due to conflicts.
+   */
+  void commit();
+
+  /**
+   * A shortcut for {@link commit} that allows users to use this catalog in try-with-resources
+   * block.
+   *
+   * @throws CommitFailedException If the updates cannot be committed due to conflicts.
+   */
+  @Override
+  default void close() {
```

Comment:
I'm not sure this is a good idea because there are times when `rollback` should be called instead. For example:

```java
try (TransactionCatalog txnCat = ...) {
  data = readData(txnCat, srcTable);
  writeData(txnCat, destTable, data);
  deleteData(txnCat, srcTable, data);
}
```

If `writeData` succeeds, then `deleteData` fails with an OOM before it creates the delete operation, then closing would complete the write without knowing that the delete didn't finish and should be rolled back. I think the fix is to accept a flag, although I'm not sure we'd want to do this:

```java
TransactionCatalog txnCat = ...
boolean rollback = true;
try {
  data = readData(txnCat, srcTable);
  writeData(txnCat, destTable, data);
  deleteData(txnCat, srcTable, data);
  rollback = false;
} finally {
  txnCat.close(rollback);
}
```

Reply:
I think the question is whether a failed commit does auto-rollback. I was thinking that it would. In that case, are you still concerned?

Reply:
With separate operations, the commit doesn't necessarily know that it failed. Say `deleteData` starts with a call to a logger, and that hits a null-pointer exception. To the transaction, that looks like a successful read, a successful write, and then a call to `commit` right?

Reply:
I'm confused by this comment. Can you expound? 

The model I'm describing is that nested transactions each have to validate or fail cleanly. If they fail, they are noops and shouldn't influence the catalog level transaction. If they validate, they are added as a pending operation to the global transaction. When the global transaction commits, either all the validated pending operations complete or they all fail.

Reply:
Here's an example case:

```java
String tableName = null;
try (TransactionCatalog txnCat = ...) {
  data = readData(txnCat, srcTable);
  writeData(txnCat, destTable, data);
  LOG.info("Cleaning up {}", tableName.split(".")[3]);
  deleteData(txnCat, srcTable, data);
}
```

The reason why `tableName` is null isn't important, it's just some bug. This code writes, then logs that it is cleaning up. But `tableName.split` throws a `NullPointerException` because the name is null, which causes the code to completely skip `deleteData`. It doesn't create the operation that should be part of the transaction.

The problem is that when the `try` block exits, it calls `commit` that has no way to know that part of the transaction never happened. All of the operations succeeded; one is just missing.

Reply:
Got it, good point. 

I implemented [RollbackCloseable](https://github.com/dremio/dremio-oss/blob/master/common/src/main/java/com/dremio/common/AutoCloseables.java#L176) to try to better handle this kind of pattern before. I'm inclined to actually use that pattern here as well. Basically, close is actually a call to rollback if commit hasn't been actively called.

Reply:
+1 for RollbackCloseable

Reply:
Updated to use rollback closeable pattern.